### DISCUSSION TITLE
Nix flake lockfile maintenance

### DISCUSSION THREAD

**Discussion by ck3mp3r** [renovatebot/renovate#24655] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

v39.0.5

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

I've successfully used nix lockfile maintenance on projects that only have a flake.lock file. However, I have a project with not only a flake.lock but also a cargo.lock file. 

In the supplied logs you can clearly see it is picking up the flake etc, but then it never acts on it. The same settings work in another project where there isn't a cargo.lock present.

The renovate config can be found at https://github.com/ck3mp3r/rmx-cli/blob/main/.github/workflows/renovate.json with the action files in the same directory.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Applying enabledManagers filtering (repository=ck3mp3r/rmx-cli)
DEBUG: Using file match: (^|/)Cargo\.toml$ for manager cargo (repository=ck3mp3r/rmx-cli)
DEBUG: Using file match: ^(workflow-templates|\.(?:github|gitea|forgejo)/workflows)/[^/]+\.ya?ml$ for manager github-actions (repository=ck3mp3r/rmx-cli)
DEBUG: Using file match: (^|/)action\.ya?ml$ for manager github-actions (repository=ck3mp3r/rmx-cli)
DEBUG: Using file match: (^|/)flake\.nix$ for manager nix (repository=ck3mp3r/rmx-cli)
DEBUG: Matched 1 file(s) for manager cargo: Cargo.toml (repository=ck3mp3r/rmx-cli)
DEBUG: Matched 3 file(s) for manager github-actions: .github/workflows/release.yaml, .github/workflows/renovate.yaml, .github/workflows/test.yaml (repository=ck3mp3r/rmx-cli)
DEBUG: Matched 1 file(s) for manager nix: flake.nix (repository=ck3mp3r/rmx-cli)
DEBUG: Neither .cargo/config nor .cargo/config.toml found (repository=ck3mp3r/rmx-cli)
DEBUG: manager extract durations (ms) (repository=ck3mp3r/rmx-cli)
       "managers": {"nix": 11, "cargo": 15, "github-actions": 16}
DEBUG: Found cargo package files (repository=ck3mp3r/rmx-cli)
DEBUG: Found github-actions package files (repository=ck3mp3r/rmx-cli)
DEBUG: Found 4 package file(s) (repository=ck3mp3r/rmx-cli)
DEBUG: Manager explicitly enabled in "enabledManagers" config, but found no results. Possible config error? (repository=ck3mp3r/rmx-cli)
       "manager": "nix"
 INFO: Dependency extraction complete (repository=ck3mp3r/rmx-cli, baseBranch=main)
       "stats": {
         "managers": {
           "cargo": {"fileCount": 1, "depCount": 9},
           "github-actions": {"fileCount": 3, "depCount": 19}
         },
         "total": {"fileCount": 4, "depCount": 28}
       }
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment]
It doesn't detect any dependencies in https://github.com/ck3mp3r/rmx-cli/blob/main/flake.nix so skips the file. I guess you're still wanting lock file maintenance to work?

**Comment by ck3mp3r** [comment]
It should treat inputs as dependencies as they are also frozen in the lockfile.

**Reply by rarkins** [reply]
I mean any _direct_ dependencies.

**Reply by ck3mp3r** [reply]
If adding a direct dependency will cause the lockfile maintenance to work I'll use that as a workaround for now...

**Reply by rarkins** [reply]
It might do, but I don't think we should make it a requirement. The issue is that Renovate is skipping the nix package file if it doesn't find dependencies

**Reply by ck3mp3r** [reply]
As a Nix user my assumption was actually that it would also keep the inputs in check.

**Reply by rarkins** [reply]
There are no versions of hashes in flake.nix for it to keep updated

**Reply by ck3mp3r** [reply]
Nix flakes don't need versions for inputs. The lock file is used to pin those to specific commits. So if I want to pin it to the latest and greatest I run a `nix flake update` and it updates the flake.lock file for me.

**Reply by rarkins** [reply]
Could you identify which lines contain the package (names) you're referring to?

**Reply by ck3mp3r** [reply]
https://github.com/ck3mp3r/rmx-cli/blob/011c9471262b26d5784a76a2d28df5b896efa947/flake.nix#L4 and the following 2 lines, which are then locked in https://github.com/ck3mp3r/rmx-cli/blob/011c9471262b26d5784a76a2d28df5b896efa947/flake.lock#L3, https://github.com/ck3mp3r/rmx-cli/blob/011c9471262b26d5784a76a2d28df5b896efa947/flake.lock#L22 and https://github.com/ck3mp3r/rmx-cli/blob/011c9471262b26d5784a76a2d28df5b896efa947/flake.lock#L40

**Reply by rarkins** [reply]
Thanks. I'm not familiar with the hows or whys of our existing parser but it seems reasonable to _not_ skip the file if it contains those.

On the other hand, I think you _gain_ visibility if you include versions in the flake.nix and let Renovate update them in addition to the lock file

**Reply by ck3mp3r** [reply]
You could add hard coded refs, but then again why would you do that if they are already pinned in the lock file...

In any case, I'll try with a dependency, alternatively I'll just run a scheduled nix flake update.

**Reply by rarkins** [reply]
The main reason is developer visibility. Generally lock files are not easy or designed for human consumption while package files are. It also provides more accuracy for updates from Renovate (it can pick an exact version to update to)

**Reply by ck3mp3r** [reply]
I'll have to resort to just using `nix flake update` to keep the lock file fresh. Inputs don't have versions (unless you count pointing at a release branch), at least not that I have come across, they are pinned to git refs via the flake.lock.

**Comment by rarkins** [comment, marked as answer]
Does this PR look accurate? https://github.com/renovate-reproductions/24655/pull/2/files

**Reply by ck3mp3r** [reply]
Yup, that looks about right. 👍🏼

**Reply by ck3mp3r** [reply]
Thanks @rarkins, much appreciated, it is working now.
